event-store/src/adapters/pg/cache.rs

```diff
@@ -20,54 +21,58 @@ impl PgCacheAdapter {
     }
 }
 
-impl<'a> CacheAdapter<PgQuery<'a>> for PgCacheAdapter {
-    fn insert<V>(&self, key: &PgQuery, value: V)
-    where
-        V: Serialize,
-    {
-        let args_hash = Sha256::digest(format!("{:?}:[{}]", key.args, key.query).as_bytes());
-
-        self.conn
-            .get()
-            .expect("Could not get PG connection")
-            .execute(
-                r#"INSERT INTO aggregate_cache (id, data, time)
+impl CacheAdapter for PgCacheAdapter {
+    fn set<'a, V: Serialize + Send + 'a>(
+        &self,
+        key: String,
+        value: V,
+    ) -> BoxedFuture<'a, (), String> {
+        let conn = self.conn.clone();
+        Box::new(FutLazy(move || {
+            conn.get()
+                .expect("Could not get PG connection")
+                .execute(
+                    r#"INSERT INTO aggregate_cache (id, data, time)
                 VALUES ($1, $2, NOW())
                 ON CONFLICT (id)
                 DO UPDATE SET data = EXCLUDED.data, time = now() RETURNING data"#,
-                &[&args_hash.as_slice(), &to_value(value).expect("To value")],
-            ).expect("Cache");
+                    &[&key, &to_value(value).expect("To value")],
+                ).expect("Update cache");
+            FutOk(())
+        }))
     }
 
-    fn get<T>(&self, key: &PgQuery) -> Option<CacheResult<T>>
+    fn get<'a, T>(&self, key: String) -> BoxedFuture<'a, Option<CacheResult<T>>, String>
     where
-        T: for<'de> Deserialize<'de>,
+        T: DeserializeOwned + Send + 'a,
     {
-        let args_hash = Sha256::digest(format!("{:?}:[{}]", key.args, key.query).as_bytes());
-
-        let rows = self
-            .conn
-            .get()
-            .expect("Could not get PG connection")
-            .query(
-                "SELECT data, time FROM aggregate_cache WHERE id = $1 LIMIT 1",
-                &[&args_hash.as_slice()],
-            ).expect("Ret");
+        let conn = self.conn.clone();
+        Box::new(FutLazy(move || {
+            let rows = conn
+                .get()
+                .expect("Could not get PG connection")
+                .query(
+                    "SELECT data, time FROM aggregate_cache WHERE id = $1 LIMIT 1",
+                    &[&key],
+                ).expect("Retrieve cache");
 
-        // `rows.get()` panics if index is out of bounds, hence this check
-        if rows.len() != 1 {
-            None
-        } else {
-            let row = rows.get(0);
+            // `rows.get()` panics if index is out of bounds, hence this check
+            if rows.len() != 1 {
+                Box::new(FutOk(None))
```

Comment:
Does this need to be boxed? - it's inside `Box::new(LazyFuture(|| {...}))` so does it need to be boxed again?

Reply:
Read again